### PR TITLE
[BEAM-717] Converts KafkaIO builders to @AutoValue

### PR DESCRIPTION
**Pull request by jkff**
This is in the same spirit as https://github.com/apache/incubator-beam/pull/1031, https://github.com/jbonofre/incubator-beam/pull/1 and https://github.com/apache/incubator-beam/pull/1033 . Semantics is unchanged AFAICT. The only user-visible change is that TypedRead and TypedWrite no longer exist (they were unnecessary in the first place) - see the trivial changes in test.

R: @rangadi

### CODE REVIEW

File: sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java [hunk 1]
```diff
@@ -268,26 +300,24 @@
...
     }
 
     /**
      * Returns a new {@link Read} with {@link Coder} for key bytes.
      */
+    @SuppressWarnings({"unchecked", "rawtypes"})
     public <KeyT> Read<KeyT, V> withKeyCoder(Coder<KeyT> keyCoder) {
-      return new Read<KeyT, V>(topics, topicPartitions, keyCoder, valueCoder,
-          consumerFactoryFn, consumerConfig, maxNumRecords, maxReadTime);
+      return ((Builder) toBuilder()).setKeyCoder(keyCoder).build();
```
Comment (rangadi) [inline sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java]: casting this way might be dangerous. See comment for withTimestampFn2()

File: sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java [hunk 2]
```diff
@@ -329,100 +354,32 @@
...
-   * A {@link PTransform} to read from Kafka topics. See {@link KafkaIO} for more
-   * information on usage and configuration.
-   */
-  public static class TypedRead<K, V>
-                      extends PTransform<PBegin, PCollection<KafkaRecord<K, V>>> {
-
-    /**
      * A function to assign a timestamp to a record. Default is processing timestamp.
      */
-    public TypedRead<K, V> withTimestampFn2(
```
Comment (rangadi) [inline sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java]: These timestamp and watermark function are the reason why we have TypedRead. We don't want to change the type once one of these methods are called. i.e, we can't have:
```
Read()
  .withKeyCoder(StringCoder())          // Read<String, byte[]>
  .withTimestampFn(function_expecting_string_key)
  .withKeyCoder(IntCoder())               // Read<Integer, byte[]>!!!
```

Comment (rangadi) [inline sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java]: Arguably the above usage is bit contrived. If we could detect it at runtime, it might be safe enough. i.e. we ensure, but still a bit messy (it could be as simple as ensuring withKeyCoder() or withValueCoder() are not invoked after setting timestamp functions).

File: sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java [hunk 3]
```diff
@@ -1076,8 +991,8 @@ public Instant getWatermark() {
         return initialWatermark;
       }
 
-      return source.watermarkFn.isPresent()
-          ? source.watermarkFn.get().apply(curRecord) : curTimestamp;
+      return source.spec.getWatermarkFn() == null
```
Comment (rangadi) [inline sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java]: != null

File: sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java [hunk 4]
```diff
@@ -1179,55 +1114,40 @@ public void close() throws IOException {
...
-      return new Write<K, V>(topic, keyCoder, valueCoder, producerConfig);
+      return toBuilder().setTopic(topic).build();
     }
 
     /**
      * Returns a new {@link Write} with {@link Coder} for serializing key (if any) to bytes.
      * A key is optional while writing to Kafka. Note when a key is set, its hash is used to
      * determine partition in Kafka (see {@link ProducerRecord} for more details).
      */
+    @SuppressWarnings({"unchecked", "rawtypes"})
```
Comment (rangadi) [inline sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java]: As we discussed, please check with Dan about requiring user to set the type initially (```KafkaIO.<String, byte[]>Read()...```).

File: sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java [hunk 5]
```diff
@@ -1179,55 +1114,40 @@ public void close() throws IOException {
...
-   * information on usage and configuration.
-   */
-  public static class TypedWrite<K, V> extends PTransform<PCollection<KV<K, V>>, PDone> {
 
     /**
      * Returns a new {@link Write} with a custom function to create Kafka producer. Primarily used
      * for tests. Default is {@link KafkaProducer}
      */
-    public TypedWrite<K, V> withProducerFactoryFn(
+    public Write<K, V> withProducerFactoryFn(
```
Comment (rangadi) [inline sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java]: same issue as with timestamp function for Read().

File: sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java [hunk 6]
```diff
@@ -224,20 +224,53 @@
...
-   * A {@link PTransform} to read from Kafka topics. See {@link KafkaIO} for more
-   * information on usage and configuration.
+   * A {@link PTransform} to read from Kafka topics. See {@link KafkaIO} for more information on
+   * usage and configuration.
    */
-  public static class Read<K, V> extends TypedRead<K, V> {
+  @AutoValue
+  public abstract static class Read<K, V>
+      extends PTransform<PBegin, PCollection<KafkaRecord<K, V>>> {
+    abstract Map<String, Object> getConsumerConfig();
```
Comment (dhalperi) [inline sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java]: presumably these need to be public?

**Comment by jkff** [comment]
OK, I changed the PR to specify types when creating the IO. PTAL.

File: sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java (hunk 6 above)
Comment (jkff) [inline sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java]: They weren't before this PR; don't see a need to expand the API surface as part of this refactoring.

**Comment by rangadi** [comment]
LGTM!

**Comment by jkff** [comment]
R: @dhalperi as committer.

**Comment by jkff** [comment]
Rebased onto checkstyle fixes.

File: sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java [hunk 7]
```diff
@@ -199,45 +200,106 @@
...
+        .setMaxNumRecords(Long.MAX_VALUE)
+        .build();
+  }
+
+  /**
+   * Creates an uninitialized {@link Write} {@link PTransform}. Before use, Kafka configuration
+   * should be set with {@link Write#withBootstrapServers(String)} and {@link Write#withTopic}
+   * along with {@link Coder}s for (optional) key and values.
+   */
+  public static Write<byte[], byte[]> writeBytes() {
```
Comment (dhalperi) [inline sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java]: I think these changes are a strict degradation in usability. Why is the `byte[]` version no longer available "for free" and instead a special builder?

Comment (jkff) [inline sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java]: I looked at usages of KafkaIO on github, and nobody was using the `byte[]` version directly - people were always overriding the key and value coders. I think this version should not have existed in the first place - or more like: if we started with `KafkaIO.<KeyT, ValueT>write()`, we probably would not have added `writeBytes()`, and wouldn't have regretted that it doesn't default to bytes.

It seems acceptable to me to require specifying how to encode/decode the received data; if the user wants to deal with raw bytes directly, that's easy to do - just use `writeBytes()` (and the case when you want to have a raw key but encode the value, or vice versa, I think, is sufficiently marginal that a user can specify the ByteArrayCoder directly).

I agree having "default values" for the type parameters and, accordingly, some fields depending on them (such as coders) can be a minor convenience, but it's very hard to do in a type-safe way in general, and I think the community benefits of having an important connector using a recommended, simple-to-code, easily-reusable pattern outweigh this cost.

Comment (jkff) [inline sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java]: Ping? (the Travis run seems to have flaked out for unrelated reasons)

**Comment by dhalperi** [comment]
IMO the AutoValue change may be an improvement but the way we had to change the usability of transforms with up-front types is a regression. But, we've not reached consensus (because I've been neglectful; sorry). About to be traveling, so handing off to another committer before I leave.

R: -@dhalperi @davorbonaci

**Comment by jkff** [comment]
I talked to @rangadi and he says that there was no practical reason for having the `byte[]` key/value type defaults (i.e. it's an artifact of the original implementation, rather than a convenience feature; and even if we wanted to have defaults for these types, it'd better have been `String`). This matches my finding that no code uses those defaults as-is. So that part should not be considered a regression.

**Comment by jkff** [comment]
retest this please

**Comment by jkff** [comment]
retest this please

File: sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java [hunk 8]
```diff
@@ -185,11 +185,11 @@
...
  * <pre>{@code
  *  PCollection<String> strings = ...;
- *  strings.apply(KafkaIO.write()
+ *  strings.apply(KafkaIO.<Void, String>write()
  *      .withBootstrapServers("broker_1:9092,broker_2:9092")
  *      .withTopic("results")
  *      .withValueCoder(StringUtf8Coder.of()) // just need coder for value
- *      .values() // writes values to Kafka with default key
+ *      .values());
  *    );
```
Comment (dhalperi) [inline sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java]: drop the `);` on the next line?

**Comment by dhalperi** [comment]
LGTM. If you prefer, I can fixup the one javadoc issue in merge; if you don't push I'll do so tomorrow.

File: sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java (hunk 8 above)
Comment (jkff) [inline sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java]: Done.

**Comment by jkff** [comment]
Done, and I squashed the commits.

**Comment by jkff** [comment]
The coveralls change seems weird. My change does not affect any of the files in which it reported changed coverage.

File: sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java [hunk 9]
```diff
@@ -116,17 +117,16 @@
...
  *  pipeline
- *    .apply(KafkaIO.read()
+ *    .apply(KafkaIO.<Long, String>read()
  *       .withBootstrapServers("broker_1:9092,broker_2:9092")
  *       .withTopics(ImmutableList.of("topic_a", "topic_b"))
- *       // above two are required configuration. returns PCollection<KafkaRecord<byte[], byte[]>
+ *       // set a Coder for Key and Value
+ *       .withKeyCoder(BigEndianLongCoder.of())
+ *       .withValueCoder(StringUtf8Coder.of())
+ *       // above four are required configuration. returns PCollection<KV<Long, String>>
```
Comment (jkff) [inline sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java]: Done (changed to KafkaRecord).

File: sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java [hunk 10]
```diff
@@ -256,10 +303,9 @@
      * of how the partitions are distributed among the splits.
      */
     public Read<K, V> withTopics(List<String> topics) {
```
Comment (jkff) [inline sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java]: This seems just as easy to do in a separate PR, since it's unrelated to the current change.

**Comment by dhalperi** [comment]
retest this please
